Google Cloud: BigQuery DataSet and Table resources

coffeepac opened the pull request:
BigQuery is a managed database-like service.  It supports SQL as
a query language and nearly infinite scale.  This PR adds support
to manage tables and datasets.

coffeepac commented:
@lwander @sparkprime anything I can do for this one or is it in the hopper, waiting its turn?

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,87 @@
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
```
It looks like you're missing quite a few features described [here](https://cloud.google.com/bigquery/docs/reference/v2/datasets)

coffeepac replied:
I am missing two general classes of things:
 - the role access information, which should be included.  I'll do that
 - updating/patching an existing dataset.  Update doesn't need to be implemented at all, in terraform's context that's handled by delete/create when a ForceNow config val is set.  I will use patch to implement Update

Anything else you'd like from that page?  I see list, which isn't useful in a terraform context, and a bunch of computed fields that don't have any obvious use in a terraform context.

lwander replied:
For starters
* `datasetReference` is described as being required. This resource as is shouldn't work as is.

Otherwise,
* `access[]`
* `friendlyName` (which you've called `name`)
* `description`
* `defaultTableExpirationMs`
* `location`
are optional, but nice to have.

When it comes to exported attributes
* `self_link` is important to include as `computed`, since other resources often refer to one another by GCE assigned URI.
* `id` for similar reasons.
* `lastModifiedTime` can be useful for people examining infrastructure managed by Terraform.

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,87 @@
+		return err
+	}
+	
+	err = resourceBigQueryDatasetRead(d, meta)
```
You can just write `return resourceBigQueryDatasetRead(d, meta)` for conciseness.

coffeepac replied:
Will do.

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,87 @@
+	config := meta.(*Config)
+	
+	call := config.clientBigQuery.Datasets.Get(config.Project, d.Get("name").(string))
+	res, err := call.Do()
```
There's some computed fields like `id` that you can set here.

coffeepac replied:
There are but I don't know if they're useful.  I don't think I need to use them for implementing patch as terraform update and otherwise I've already been asked to remove computed fields that are unused on a different PR.  Is there a reason for any of the computed fields?

lwander replied:
Which provider was that PR for? I've always been on the side of keeping computed attributes, since they can be useful for people examining with `terraform show` (unless they are totally useless during inspection like a fingerprint hash, or a list of errors that should've surfaced already).

coffeepac replied:
https://github.com/hashicorp/terraform/pull/3671  a comment from SparkPrime regarding the Topic's canonical name.  This might be a different context as the Topic name may be totally useless outside of creation time.

lwander replied:
Ah, I think he made that comment because `cannonical_name` doesn't appear with that name in the [JSON API](https://cloud.google.com/pubsub/reference/rest/v1/projects.subscriptions/create). We are more concerned with keeping the schema format as close as possible to the JSON that's sent to the GCP endpoint.

coffeepac replied:
got it.  I'll chuck 'em in then.  adding computed fields is easy.

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func resourceBigQueryDatasetRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
```
Can you match the read behavior I implemented in #3913?

coffeepac replied:
Yes!  I saw that PR, not sure why I didn't do it here.

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,87 @@
+	config := meta.(*Config)
+
+	if d.Get("can_delete").(bool) == true {
+		call := config.clientBigQuery.Datasets.Delete(config.Project, d.Get("name").(string))
```
`x == true` is redundant. 

Also, why do you have this flag? I'm not clear why we need to specify `can_delete` here when it's not present in any other resource.

coffeepac replied:
hah, right.  will clean that up.

can_delete:  I wanted an extra guard on datasets/tables because I can easily see, and I have some usecases for, the created data outliving the system that created it.  So I would want to destroy everything in a terraform deploy except the outputs.  And at that point, I don't want terraform involved anymore.  The manual way to do this is remove the resource from the config file and from the state file, but that's not something I relish automating.

Is there any concept in terraform that encompasses the idea of 'remove this resource from the statefile but don't delete it'.  That's what I'm trying to get at somewhat awkwardly with can_delete.

lwander replied:
Ah I see what you're doing. I don't think that will work, since `d.Get("can_delete")` will query an attribute that's missing and default to `false`. 

What you're looking for is already built in [here](https://terraform.io/docs/configuration/resources.html#prevent_destroy).

coffeepac replied:
that's very close to what I want but it doesn't delete the resource from terraform's statefile, it means I can't tear down the system that created the resource.

is there a way to run destroy that leaves some resources?  that would also work for this.

---
I will elaborate a bit to make this concrete.  I want to deploy a transient data processing system.  Transient but long lived and it will have its internals re-arragned an arbitrary number of times by terraform (hence my PRs for pubsub and dataflow).  At some point, we will want to tear the entire system down.  At that point I would like to be able to clear all of my terraform resources out except for my created datasets.  I believe the easiest way to do that is to trick terraform into thinking it has deleted the datasets by clearing out the resource's ID.  But I am open to suggestions.

lwander replied:
Ah, interesting. I see two ways you can achieve this. 

1. Using two separate directories each with their own `.tf` file you can separate the datasets from the rest of the infrastructure. This way running `terraform destroy` will only delete the infrastructure in the directory you are applying it to. The downside here is that you can't do any special interpolation syntax on resources between directories... however you could move all your top level names and definitions into a `var.tf` file, and create symlinks to it in each directory, and then reference the contents of that in each `.tf` file.

2. If you're OK with tricking Terraform you can simply remove the dataset resource from both the `.tf` file, and the `terraform.tfstate` file, causing Terraform to "forget" that it ever managed that resource. Then running `terraform destroy` should keep that dataset intact.

coffeepac replied:
Interesting, I hadn't considered #1 before.  The terraform configs are going to be autogenerated by something else which is something like a query planner so putting them in two directories and putting the prevent_destroy flag on the datasets would work.

Okay, I'll pull the can_delete flag and work the safety for the dataset into the config file generator.

lwander replied:
Awesome, good luck!

BTW, what are you using for your config file generator? @sparkprime has written a cool tool called https://github.com/google/jsonnet/ which he uses to generate Terraform configs.

coffeepac replied:
I'm not sure!  the generation is coming from jvm land, which is someone else's job at this point.  I'll check out jsonnet for future work though. thanks!

coffeepac commented:
@lwander Thanks man!  I really appreciate you taking the time to review!

lwander commented:
@coffeepac you got it! Thanks for doing all this work so far.

coffeepac commented:
@lwander I wanted to get your thoughts on my changes to the Dataset resource before I do something similar to the table and then go implement the Patch operation.

The only comment I didn't directly implement was the comment about 'datasetReference'.  This is a struct composed of datasetId and projectId which I'm pulling from the provider configuration. If that's wrong/inconsistent lemme know and I'll rework it.

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,235 @@
+				Optional: true,
+			},
+
+			"access": &schema.Schema{
```
This gets a bit tricky, since we implemented ACLs for GCS a little differently, but that was only possible because the API is different. The idea there was, you just provide a list of role entity pairs like [this](https://www.terraform.io/docs/providers/google/r/storage_bucket_acl.html). The GCS API basically squashed `userByEmail`, `groupByEmail`, and `domain` all into one field, so all we had to do was have the user concatenate the contents of that field with the `role`, and that would be the access control object. Second, there was no `view` type object attached with each GCS ACL, since the ACL is an entirely separate resource for GCS objects and buckets. Basically, to match the GCS Terraform interface, we'd really have to contort the BigQuery API, which is a shame. 

I'd say what you have here is best, I'm just leaving this comment in case people have questions about this decision in the future. 

@sparkprime, what do you think?

lwander commented on builtin/providers/google/resource_bigquery_dataset.go:
```diff
@@ -0,0 +1,235 @@
+	return nil
+}
+
+func resourceBigQueryDatasetUpdate(d *schema.ResourceData, meta interface{}) error {
```
A note about the updates, the Access Controls should be managed like the Instances I talk about [here](https://github.com/hashicorp/terraform/pull/4087#discussion_r46320543). Let me know if my explanation there is unclear. 

Also, you mentioned patch semantics. They can get a little tricky, since `null, "", 0` values are treated as "don't update", so you need to explicitly provide the API with which fields are being patched.

coffeepac replied:
@lwander thanks for the link to handle Access Controls gracefully.  I will endeavor to do that.

I'm digging into the patch for updates now.  Seems sane if a bit verbose to iterate through all optional HasChanges and then GetChange and only update the local copy of things tf has known about in the recent past, right?

I think the process would be:
- get all changes via the prior paragraph
- shove those changes into a new Dataset struct
- pass that struct into a Patch call off of the DatasetService
- call Do

I can see myself going bonkers typing all of that but seems straightforward.  Am I missing anything, big picture?  And I'm not seeing any tests of update functions, is there any you'd like me to include?

thanks for your patience with me being so slow on this.

lwander commented:
When it comes to the `datasetReference`, you're totally right for dropping `projectId`, so what you have looks good.

Are you planning on flushing out the table resource as well? It has quite a few fields [here](https://cloud.google.com/bigquery/docs/reference/v2/tables).

coffeepac commented:
I am planning on fleshing out the table resource as well. I wanted some
feedback before I did it though. Make sure I'm in the right path. Thanks
for the feedback!
On Dec 4, 2015 6:33 AM, "Lars Wander" <notifications@github.com> wrote:

> When it comes to the datasetReference, you're totally right for dropping
> projectId, so what you have looks good.
>
> Are you planning on flushing out the table resource as well? It has quite
> a few fields here
> <https://cloud.google.com/bigquery/docs/reference/v2/tables>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/terraform/pull/3764#issuecomment-161981129>.
>

coffeepac commented:
@lwander oh holidays...  I haven't started working on patch semantics yet but I have added bigquery table.  tables can have an arbitrary depth of record fields and I didn't see a clean way of doing that with helper/schema so I fudged it by having the schema in code allow for depth of one but I added a schemaFile that expects json and allows for arbitrary depth.  

the code to support arbitrary depth of record fields is there so if there's a clean way to add them into the schema I'm all for it if you have a way.  

either way, I like having the schema in a separate file so if there is a 4 layer deep table (my current use case) it isn't garbaging up the tf file.  sound okay or is that no go?

lwander commented:
Could you point me to the `schemaFile` attribute? Since Terraform doesn't have an expressive enough syntax `schemaFile` sounds like a much better solution.

coffeepac commented:
line 46 in builtin/providers/google/resource_bigquery_table.go

test using this file is at line 29 and the resource is defined at 105 in builtin/providers/google/resource_bigquery_table_test.go

the test fixture for this is:  builtin/providers/google/test-fixtures/fake_bigquery_table.json

the test schemaFile goes three levels deep.

heimweh commented:
@coffeepac I'd love to see support for BigQuery in Terraform.
I can see that this PR hasn't been updated in a while, 
do you plan to continue your work on this or do you mind if I take a stab at this? :)

coffeepac commented:
@heimweh feel free to take it over!  I got pulled off and had to create an external plugin for this.  its here:  https://github.com/22Acacia/terraform-provider-googlebigquery  Its currently in production and does what we need so we're just living with it for now.  

The big chunk of work I never got to was how to implement Update, its currently a big ball of lying No-Op.  The golang docs for the bigquery library have taken a massive leap forward in the past year so it shouldn't be a slog anymore.  It was basically just the auto-generated go doc stuff.  

Lemme know if you want me to review anything, open up permissions on a branch somewhere or have more questions.  I'm happy to help!

kurtmaile commented:
Hi all,  Any plans to provide a resource for bigquery views?  Would be super helpful!!

Thanks for your hard work on these core resources.....

kurtmaile commented:
Hi all,  Any plans to provide a resource for bigquery views?  Would be super helpful!!

Thanks for your hard work on these core resources.....

coffeepac commented:
this was merged elsewhere.  going to close this PR.
